### PR TITLE
Kernel Memory Updates (Part 7): Various fixes to code memory (Skyline support)

### PR DESCRIPTION
**Pull request by bunnei**
This is a continuation of my work to bring our kernel memory management up to speed with recent system updates and the latest known information (see previous PRs #8013, https://github.com/yuzu-emu/yuzu/pull/7974, https://github.com/yuzu-emu/yuzu/pull/7956, https://github.com/yuzu-emu/yuzu/pull/7701, https://github.com/yuzu-emu/yuzu/pull/7698, and https://github.com/yuzu-emu/yuzu/pull/7684 for a history of this effort).

With this change, we:
* Fix implementations of LockForCodeMemory & UnlockForCodeMemory.
* Fix MapProcessMemory bug where we were mapping onto virtual address space.
* Fix KCodeMemoryBug where we were addressing physical memory with a virtual address.
* Fix a bad alignment check in UnmapProcessCodeMemory.
* Various cleanup & smaller improvements.

The goal with this batch of changes is to wrap up the work from #7392 and get [Skyline](https://github.com/skyline-dev/skyline) working in yuzu. Things should be working better now, but I am not too familiar with Skyline, so please report any remaining issues here.

Thanks to @jam1garner, @itsmeft24 and @tech-ticks for assistance here and for doing most of the legwork.

### CODE REVIEW

**Comment by Struggleton** [comment]
When loading Skyline as a LayeredExeFS mod, Smash Ultimate seems to error out with error code (2069-0000) then after clicking OK on the error message, it hangs on the shader loading screen. Closing Yuzu after this crash leaves the process behind in Task Manager, even after stopping the emulation. (I'm not knowledgeable about the code base but possibly b/c qtwebengine is not being closed after emulation?)

image

Attached is a log of the crash + the skyline mod folder used to produce the crash (the Skyline folder should be placed under yuzu/load/01006A800016E000.)

[Skyline.zip](https://github.com/yuzu-emu/yuzu/files/8356016/Skyline.zip)
[yuzu_log.txt](https://github.com/yuzu-emu/yuzu/files/8356020/yuzu_log.txt)

File: src/core/hle/kernel/k_page_table.h [hunk 1]
```diff
@@ -159,10 +164,37 @@ class KPageTable final {
...
+                            KMemoryPermission new_perm, KMemoryAttribute lock_attr,
+                            const KPageLinkedList* pg);
+
+    ResultCode MakePageGroup(KPageLinkedList& pg, VAddr addr, size_t num_pages);
+
     bool IsLockedByCurrentThread() const {
         return general_lock.IsLockedByCurrentThread();
     }
 
+    bool IsHeapPhysicalAddress(const KMemoryLayout& layout, PAddr phys_addr) {
```
Comment (lioncash) [inline src/core/hle/kernel/k_page_table.h:184]:
```suggestion
    bool IsHeapPhysicalAddress(const KMemoryLayout& layout, PAddr phys_addr) const {
```

**Comment by bunnei** [comment]
@Struggleton should be fixed in next EA. Thanks!

**Comment by Struggleton** [comment]
Tried it, and Skyline by itself seems to load just fine - however when I tried using [Arcropolis ](https://github.com/Raytwo/ARCropolis)(a file replacement/modloader for Smash that uses Skyline) along with it the application hangs on the shader loading screen and memory usage just starts going up seems to spam the log with unmapped reads and unimplemented instruction @ 0x7FDC000 

I apologize for not including Arcropolis the first time around as a practical example of a Skyline use case. I've attached a log after it tries to load Arcropolis and the Skyline + Arcropolis mod I am using to produce this hang.  (the Skyline+Arcropolis folder should be placed under yuzu/load/01006A800016E000. You should be able to delete the other Skyline folder from the previous test)

[yuzu_log.zip](https://github.com/yuzu-emu/yuzu/files/8356427/yuzu_log.zip)
[Skyline+Arcropolis.zip](https://github.com/yuzu-emu/yuzu/files/8356432/Skyline%2BArcropolis.zip)

File: src/core/hle/kernel/k_page_table.h (hunk 1 above)
Comment (bunnei) [reply src/core/hle/kernel/k_page_table.h:184]: By design, KMemoryLayout caches the region found into cached_physical_heap_region

**Comment by bunnei** [comment]
@Struggleton looks like the issues with Acropolis are related to file system, and out of scope for the kernel/memory fixes intended with this PR. We'll look into it with a follow-up change.

**Pull request merged**